compiler/src/lnntoamm/Types.ts

```diff
@@ -197,83 +252,426 @@ class Struct extends Type {
+        // TODO: this might change depending on the opaque types
+        // we introduce
+        return false;
+      } else if (ty instanceof Struct) {
+        // ty is a Struct, ensure it has the same fields
+        // if (!this.fields.every(field => ty.fields[field.name] && ty.fields[field.name].eq(field.ty))) {
+        //   return false;
+        // }
+        if (!this.fields.every(field => Has.field(field, ty))) {
+          return false;
+        }
+      }
+
+      // check methods
+      return this.methods.every(m => Has.method(m, scope, ty)) &&
+        this.operators.every(o => Has.operator(o, scope, ty));
+      // check operators
+    } else if (ty instanceof Interface) {
+      // ensure `ty ⊆ this`
+      return ty.fields.every(field => this.fields.find(f => f.eq(field))) &&
+        ty.methods.every(method => this.methods.find(m => m.eq(method))) &&
+        ty.operators.every(operator => this.operators.find(o => o.eq(operator)));
+    } else if (ty instanceof HasField) {
+      return this.fields.some(field => field.eq(ty));
+    } else if (ty instanceof HasOperator) {
+      return this.operators.some(operator => operator.eq(ty));
+    } else if (ty instanceof HasMethod) {
+      return this.methods.some(method => method.eq(ty));
+    } else if (ty instanceof Generated || ty instanceof OneOf) {
+      return ty.compatibleWithConstraint(this, scope);
+    } else {
+      throw new Error(`unsure of what type the constraint is - this error should never be thrown!`);
+    }
   }
 
-  constrain(to: Type) {
-    return TODO();
+  constrain(ty: Type, scope: Scope) {
+    const baseErrorString = `type ${ty.name} was constrained to interface ${this.name} but doesn't have`;
+    this.fields.forEach(f => {
+      if (!ty.compatibleWithConstraint(f, scope)) {
+        throw new Error(`${baseErrorString} field ${f.name} with type ${f.ty}`);
+      }
+    });
+    this.methods.forEach(m => {
+      if (!Has.method(m, scope, ty)) {
+        throw new Error(`${baseErrorString} method ${m.name}(${m.params.map(p => p === null ? ty : p).map(t => t.name).join(', ')})`);
+      }
+    });
+    this.operators.forEach(o => {
+      if (Has.operator(o, scope, ty)) return;
+      if (o.isPrefix) {
+        throw new Error(`${baseErrorString} prefix operator \`${o.name} ${ty.name}\``);
+      } else {
+        throw new Error(`${baseErrorString} infix operator \`${o.params[0] || ty.name} ${o.name} ${o.params[1] || ty.name}\``);
+      }
+    });
   }
 
   eq(that: Equalable): boolean {
-    // TODO: gets more complicated later
-    return that instanceof Interface && this === that;
+    if (that instanceof Generated) {
+      return that.eq(this);
+    } else if (that instanceof Interface) {
+      // FIXME: this is technically wrong, but there's no other way
+      // to get the current generic types working without depending
+      // on `eq` returning this. Ideally, we would be checking to
+      // make sure all of the constraints match
```

Comment:
I don't quite follow this comment because generics haven't been implemented yet?

Reply:
ah, the comment's wording is a little off. I mean the current way of doing generic fn parameters (all `Stringifiables` are the same)